src/core/server/integration_tests/saved_objects/migrations/zdt_2/v2_to_zdt_switch.test.ts

```diff
@@ -0,0 +1,150 @@
+      attributes: {
+        keyword: `a_${number}`,
+        boolean: true,
+      },
+    }));
+
+    await savedObjectsRepository.bulkCreate(sampleAObjs);
+
+    return { client };
+  };
+
+  describe('when switching from a compatible version', () => {
+    it('it able to re-use a cluster state from the v2 algorithm', async () => {
+      await createBaseline();
+
+      const typeA = getSampleAType();
+
+      const { runMigrations, client, savedObjectsRepository } = await getKibanaMigratorTestKit({
+        ...getBaseMigratorParams({ migrationAlgorithm: 'zdt' }),
+        logFilePath,
+        types: [typeA],
+      });
+
+      await runMigrations();
+
+      const indices = await client.indices.get({ index: '.kibana*' });
+      expect(Object.keys(indices)).toEqual([v2IndexName]);
+
+      const index = indices[v2IndexName];
+      const mappings = index.mappings ?? {};
+      const mappingMeta = mappings._meta ?? {};
+
+      expect(mappings.properties).toEqual(
+        expect.objectContaining({
+          sample_a: typeA.mappings,
+        })
+      );
+
+      expect(mappingMeta).toEqual({
+        docVersions: { sample_a: '10.1.0' },
+        mappingVersions: { sample_a: '10.1.0' },
+        migrationState: expect.any(Object),
+      });
+
+      const { saved_objects: sampleADocs } = await savedObjectsRepository.find({
+        type: 'sample_a',
+      });
+
+      expect(sampleADocs).toHaveLength(5);
+
+      const records = await parseLogFile(logFilePath);
+      expect(records).toContainLogEntries(
+        [
+          'INIT: current algo check result: v2-compatible',
+          'INIT -> UPDATE_INDEX_MAPPINGS',
+          'INDEX_STATE_UPDATE_DONE -> DOCUMENTS_UPDATE_INIT',
+          'Migration completed',
+        ],
+        { ordered: true }
+      );
+    });
+  });
+
+  describe('when switching from an incompatible version', () => {
+    it('fails and throws an explicit error', async () => {
+      const { client } = await createBaseline({ kibanaVersion: '8.7.0' });
+
+      // even when specifying an older version, the `indexTypeMap` will be present on the index's meta,
+      // so we have to manually remove it there.
+      const indices = await client.indices.get({
+        index: '.kibana_8.7.0_001',
+      });
+      const meta = indices['.kibana_8.7.0_001'].mappings!._meta! as IndexMappingMeta;
+      delete meta.indexTypesMap;
+      await client.indices.putMapping({
+        index: '.kibana_8.7.0_001',
+        _meta: meta,
+      });
+
+      const typeA = getSampleAType();
+      const { runMigrations } = await getKibanaMigratorTestKit({
+        ...getBaseMigratorParams({ migrationAlgorithm: 'zdt' }),
+        logFilePath,
+        types: [typeA],
+      });
+
+      await expect(runMigrations()).rejects.toThrowErrorMatchingInlineSnapshot(
+        `"Unable to complete saved object migrations for the [.kibana] index: Index .kibana_8.7.0_001 is using an incompatible version of the v2 algorithm"`
```

Comment:
To me, the highest risk with switching from v2 to zdt: kibana has to go through the index splitting first.